automation/scripts/clean_pipelines/clean_pipelines.py

```diff
@@ -0,0 +1,422 @@
+    experiment = kfp_client.get_experiment(experiment_id=experiment_id)
+    return experiment.name if experiment else ""
+
+
+def _delete_items(
+    context: mlrun.MLClientCtx,
+    items: list[tuple[str, str]],
+    delete_func: typing.Callable[[str], None],
+    item_type: str = "run",
+) -> None:
+    """
+    A generic function to delete items such as runs or experiments and log the results.
+
+    :param context: The context object to log results.
+    :param items: A list of tuples, where each tuple contains the item ID and name to be deleted.
+    :param delete_func: The function responsible for deleting each item.
+                        It should take an ID as its argument.
+    :param item_type: The type of items being deleted (used for logging).
+                      Defaults to "run".
+    """
+    total = len(items)
+
+    context.log_result(key=f"{item_type}s_total", value=total)
+    mlrun.utils.logger.info(f"Deleting {total} {item_type}s")
+
+    deleted, failed = _perform_deletion(items, delete_func, total, item_type)
+    _log_results(context, deleted, failed, item_type)
+
+
+def _perform_deletion(
+    items: list[tuple[str, str]],
+    delete_func: typing.Callable[[str], None],
+    total_items_amount: int,
+    item_type: str = "run",
+) -> tuple[list[tuple[str, str]], list[tuple[str, str, Exception, str]]]:
+    """
+    This function iterates through a list of items, attempts to delete each item using
+    the provided `delete_func`, and logs progress. It returns a list of successfully deleted
+    items and a list of items that failed to delete with associated error details.
+
+    :param items: A list of tuples, where each tuple contains the item ID and name to be deleted.
+    :param delete_func: A callable responsible for deleting each item, taking the item ID as an argument.
+    :param total_items_amount: The total number of items to delete, used for logging progress.
+    :param item_type: The type of items being deleted, used in logging messages (defaults to "run").
+
+    :return: A tuple containing:
+             - deleted_items: A list of tuples of successfully deleted items, with each tuple containing (name, ID).
+             - failed_items: A list of tuples for failed deletions, with each tuple containing
+               (name, ID, exception, exception reason).
+    """
+    deleted_count = 0
+    failed_items = []
+    deleted_items = []
+
+    for item_id, name in items:
+        try:
+            delete_func(item_id)
+            deleted_count += 1
+            deleted_items.append((name, item_id))
+            if deleted_count % DELETION_LOG_INTERVAL == 0:
+                mlrun.utils.logger.info(
+                    f"Deleted {deleted_count}/{total_items_amount} {item_type}s successfully"
+                )
+                # A 5-second sleep is used to balance KFP load and limit the rate of deletion requests.
+                sleep(5)
```

Comment:
I'm assuming it's possible that one of the empty experiments will get a new run while deleting all of the experiments right? Is there something we can do about it? verify no matching schedules?

Reply:
If an experiment is deleted and a new run is created afterward, the experiment will automatically be recreated. However, this script is made to be used to delete older runs, meaning that if an experiment becomes empty as a result, there have likely been no new runs in the past few weeks (based on the filtering criteria). Therefore, the likelihood of a new run being added immediately after deletion is quite low, making this scenario a real edge case.

That said, I can look into adding a verification step to check for matching schedules to further minimize this possibility, though I believe the chances of it happening is quite low.

Reply:
That's ok, worst case it will fail and continue